src/credentials_rotators/npm/lambda_functions/user_login_password_rotator.py

```diff
@@ -0,0 +1,79 @@
+import json
+import os
+
+from npm_credentials_rotator import NPMCredentialsRotator
+from npm_utils import get_user_info_using_password, update_login_password
+from secrets_config_utils import get_secret_config, get_secret_key
+from secrets_manager_utils import get_secret_value
+
+
+class UserLoginPasswordRotator(NPMCredentialsRotator):
+    """
+    Handles the rotation logic specific to rotating the NPM user login password secret
+    """
+
+    def __init__(self, arn, token, step):
+        super().__init__(arn, token, step)
+
+    def create_secret(self):
+        """Create the secret
+        This method first checks for the existence of a secret for the passed in request token. If one does not exist, it will generate a
+        new password for the NPM user and put it with the passed in request token.
+        Raises:
+            ResourceNotFoundException: If the secret with the specified arn and stage does not exist
+        """
+        # Make sure the current secret exists
+        self.check_secret_exists()
+
+        # Now try to get the secret version, if that fails, put a new secret
+        try:
+            self.service_client.get_secret_value(SecretId=self.arn, VersionId=self.token, VersionStage="AWSPENDING")
+            self.logger.info("createSecret: Successfully retrieved secret for %s." % self.arn)
+        except self.service_client.exceptions.ResourceNotFoundException:
+            # Get exclude characters from environment variable used to create new password
+            exclude_characters = os.environ[
+                'EXCLUDE_CHARACTERS'] if 'EXCLUDE_CHARACTERS' in os.environ else '/@"\'\\'
+            # Generate a random password
+            new_password = self.service_client.get_random_password(ExcludeCharacters=exclude_characters)[
+                'RandomPassword']
+            npm_login_password_secret_config = get_secret_config('npm_login_password_secret')
+            npm_login_password_secret_key = get_secret_key(npm_login_password_secret_config)
+            npm_login_password_secret = json.dumps({npm_login_password_secret_key: new_password})
+
+            # Put the secret
+            self.service_client.put_secret_value(SecretId=self.arn,
+                                                 ClientRequestToken=self.token,
+                                                 SecretString=npm_login_password_secret,
+                                                 VersionStages=['AWSPENDING'])
```

Comment:
Kind of a lot of work going on in this exception block, but arguably the exception block isn't exceptional in this use case--the point is to create a new secret.

You could better express your intent by returning from the end of the `try` block and make the `except` a very short "log" statement along the lines of "Secret not found, creating new one" or similar. Then promote the guts of your `except` clause to the main flow of the method.

Also consider logging the success "Successfully retrieved secret..." log a warning, since you probably shouldn't be calling `create_secret` if one already exists?

Reply:
Good refactor. Thanks. Also, moved the creation of new random password to a separate method.